chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -8,47 +8,160 @@
+  Member<StorageNamespace> local_storage_;
+};
+
+const char EphemeralStorageNamespaces::kSupplementName[] =
+    "EphemeralSessionStorageNamespaces";
+
+EphemeralStorageNamespaces::EphemeralStorageNamespaces(
+    StorageController* controller,
+    const String& session_storage_id,
+    const String& local_storage_id)
+    : session_storage_(
+          MakeGarbageCollected<StorageNamespace>(controller,
+                                                 session_storage_id)),
+      local_storage_(MakeGarbageCollected<StorageNamespace>(controller,
+                                                            local_storage_id)) {
+}
+
+void EphemeralStorageNamespaces::Trace(Visitor* visitor) const {
+  visitor->Trace(session_storage_);
+  visitor->Trace(local_storage_);
+  Supplement<Page>::Trace(visitor);
 }
 
-}  // namespace
+// static
+EphemeralStorageNamespaces* EphemeralStorageNamespaces::From(Page* page) {
+  if (!page)
+    return nullptr;
+
+  EphemeralStorageNamespaces* supplement =
+      Supplement<Page>::From<EphemeralStorageNamespaces>(page);
+  if (supplement)
+    return supplement;
+
+  // The ephemeral session storage namespace is constructed using the
+  // normal session storage id. We get it here and transform it.
+  StorageNamespace* session_storage_namespace = StorageNamespace::From(page);
+  if (!session_storage_namespace)
+    return nullptr;
 
-BraveDOMWindowStorage::BraveDOMWindowStorage() {}
-BraveDOMWindowStorage::~BraveDOMWindowStorage() {}
+  String session_storage_id = session_storage_namespace->namespace_id() +
+                              String("ephemeral-session-storage");
+  String local_storage_id = session_storage_namespace->namespace_id() +
+                            String("ephemeral-local-storage");
+  supplement = MakeGarbageCollected<EphemeralStorageNamespaces>(
+      StorageController::GetInstance(), session_storage_id, local_storage_id);
+
+  ProvideTo(*page, supplement);
+  return supplement;
+}
 
 // static
-StorageArea* BraveDOMWindowStorage::sessionStorage(LocalDOMWindow& window,
-                                              ExceptionState& exception_state) {
-  auto* storage =
-      DOMWindowStorage::From(window).sessionStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+const char BraveDOMWindowStorage::kSupplementName[] = "DOMWindowStorage";
+
+// static
+BraveDOMWindowStorage& BraveDOMWindowStorage::From(LocalDOMWindow& window) {
+  BraveDOMWindowStorage* supplement =
+      Supplement<LocalDOMWindow>::From<BraveDOMWindowStorage>(window);
+  if (!supplement) {
+    supplement = MakeGarbageCollected<BraveDOMWindowStorage>(window);
+    ProvideTo(window, supplement);
+  }
+  return *supplement;
 }
 
 // static
-StorageArea* BraveDOMWindowStorage::localStorage(LocalDOMWindow& window,
-                                            ExceptionState& exception_state) {
+StorageArea* BraveDOMWindowStorage::sessionStorage(
```

Comment:
how do we know whether the feature is enabled here or not?

Reply:
We don't. We had assumed that if third-party DOM was allowed at all (which is controlled now by `cookie_settings_base.cc`) then the feature was enabled. If this assumption isn't a good one, we will need to plumb through the feature to this place.

Reply:
I've modified this to use the feature now that it's available here with the most recent changes to #6615.